src/components/PopupWithForm.js

```diff
@@ -20,14 +22,48 @@ export default class PopupWithForm extends Popup {
     super.setEventListeners();
     this._form.addEventListener("submit", (event) => {
       event.preventDefault();
-      this._handleFormSubmit(this._getInputValues());
+      this._renderLoading(true); // Show "Saving..." text
+      this._handleFormSubmit(this._getInputValues())
+        .then(() => {
+          this._renderLoading(false); // Reset to original text
+          this.close(); // Close the popup after submission
+        })
+        .catch(() => {
+          this._renderLoading(false); // Ensure to reset in case of error
+        });
       this._form.reset();
```

Comment:
You need to **reset** the form only   in block `then` in order not to do that immediately after `submit` because when there is an error, the modal will still be opened and the user can try to send the data again to the server without having to type the data into the inputs once more.